sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaReadSchemaTransformProvider.java

```diff
@@ -234,6 +155,74 @@ public PCollectionRowTuple expand(PCollectionRowTuple input) {
+            "To read from Kafka in RAW format, you can't provide a schema.");
+      }
+      beamSchema = Schema.builder().addField("payload", Schema.FieldType.BYTES).build();
+      valueMapper = getRawBytesToRowFunction(beamSchema);
+    } else if (format != null && format.equals("PROTO")) {
+      String messageName = configuration.getMessageName();
+      if (messageName == null) {
+        throw new IllegalArgumentException("Expecting a messageName to be non-null.");
+      }
+      String fileDescriptorPath = configuration.getFileDescriptorPath();
+      if (fileDescriptorPath != null) {
+        beamSchema = ProtoByteUtils.getBeamSchemaFromProto(fileDescriptorPath, messageName);
+        valueMapper = ProtoByteUtils.getProtoBytesToRowFunction(fileDescriptorPath, messageName);
+      } else {
+        beamSchema = ProtoByteUtils.getBeamSchemaFromProtoSchema(inputSchema, messageName);
+        valueMapper = ProtoByteUtils.getProtoBytesToRowFromSchemaFunction(inputSchema, messageName);
+      }
+    } else {
+      beamSchema =
+          Objects.equals(format, "JSON")
+              ? JsonUtils.beamSchemaFromJsonSchema(inputSchema)
+              : AvroUtils.toBeamSchema(new org.apache.avro.Schema.Parser().parse(inputSchema));
+      valueMapper =
+          Objects.equals(format, "JSON")
+              ? JsonUtils.getJsonBytesToRowFunction(beamSchema)
+              : AvroUtils.getAvroBytesToRowFunction(beamSchema);
+    }
+    return new SchemaTransform() {
+      @Override
+      public PCollectionRowTuple expand(PCollectionRowTuple input) {
+        KafkaIO.Read<byte[], byte[]> kafkaRead =
+            KafkaIO.readBytes()
+                .withConsumerConfigUpdates(consumerConfigs)
+                .withConsumerFactoryFn(new ConsumerFactoryWithGcsTrustStores())
+                .withTopic(configuration.getTopic())
+                .withBootstrapServers(configuration.getBootstrapServers());
+        if (isTest) {
+          kafkaRead = kafkaRead.withMaxReadTime(Duration.standardSeconds(testTimeoutSecs));
+        }
+
+        PCollection<byte[]> kafkaValues =
+            input.getPipeline().apply(kafkaRead.withoutMetadata()).apply(Values.create());
+
+        Schema errorSchema = ErrorHandling.errorSchemaBytes();
+        PCollectionTuple outputTuple =
+            kafkaValues.apply(
+                ParDo.of(
+                        new ErrorFn(
+                            "Kafka-read-error-counter", valueMapper, errorSchema, handleErrors))
+                    .withOutputTags(OUTPUT_TAG, TupleTagList.of(ERROR_TAG)));
+
+        PCollectionRowTuple outputRows =
+            PCollectionRowTuple.of("output", outputTuple.get(OUTPUT_TAG).setRowSchema(beamSchema));
+
+        PCollection<Row> errorOutput = outputTuple.get(ERROR_TAG).setRowSchema(errorSchema);
+        if (handleErrors) {
+          outputRows =
+              outputRows.and(
+                  Objects.requireNonNull(configuration.getErrorHandling()).getOutput(),
```

Comment:
nit: redundant since `ErrorHandling.hasOutput(configuration.getErrorHandling());` is already called
```suggestion
                  configuration.getErrorHandling().getOutput(),
```